common/base/tests/Test_UtilsThreadsafeContainers.cpp

```diff
@@ -0,0 +1,237 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include "gtest/gtest.h"
+
+#include <claragenomics/utils/threadsafe_containers.hpp>
+
+#include <algorithm>
+#include <mutex>
+#include <cstdint>
+#include <numeric>
+#include <thread>
+
+namespace claragenomics
+{
+
+// *** test ThreadsafeDataProvider ***
+
+void test_threadsafe_data_provider(const std::size_t number_of_elements,
+                                   const std::size_t number_of_threads)
+{
+    std::vector<std::size_t> data(number_of_elements);
+    std::iota(std::begin(data),
+              std::end(data),
+              0);
+
+    ThreadsafeDataProvider<std::size_t> data_provider(std::move(data));
+
+    std::mutex occurrences_per_element_mutex; // using mutex instead of atomic as the test was failining when using more than 1'000'000 atomics
+    std::vector<std::int32_t> occurrences_per_element(number_of_elements, 0);
+
+    std::vector<std::thread> threads;
+
+    for (std::size_t thread_id = 0; thread_id < number_of_threads; ++thread_id)
+    {
+        threads.push_back(std::thread([&data_provider, &occurrences_per_element, &occurrences_per_element_mutex]() {
+            while (true)
+            {
+                cga_optional_t<std::size_t> val = data_provider.get_next_element();
+
+                if (!val) // reached the end
+                {
+                    break;
+                }
+                else
+                {
+                    std::lock_guard<std::mutex> occurrences_per_element_mutex_lock(occurrences_per_element_mutex);
+                    occurrences_per_element[val.value()]++;
+                }
+            }
+        }));
+    }
+
+    for (std::size_t thread_id = 0; thread_id < number_of_threads; ++thread_id)
```

Comment:
Optional: How about an `for (std::thread& t : threads)` loop instead?

Reply:
yes, done